### ISSUE TITLE
Option idea: NoImplicityCastAny

### ISSUE THREAD

**Issue by jargv** [Microsoft/TypeScript#4787]
Here's a use case I have in mind for typescript, at which it failed:

I'm coding along, enjoying optional types and cranking out my feature. After a while I decide that my interfaces are well-understood and I create a `DefinedThing` interface. So I change my function from `foo(thing : any) : void {...}` to `foo(thing : DefinedThing) : void {...}`. The problem is that none of my other code which uses `foo` uses types yet, so Typescript will happily convert `any` into `DefinedThing` in calling `foo`. Adding the type has given me no benefit at all.

What I would have liked to happen in this case is that typescript would give me errors anywhere that I'm trying to use an `any` in place of a `DefinedThing`. Following these errors would lead me to be much more explicit throughout my code, but only in the places where I have nailed down my interfaces. All of the rest of my code would continue to be untyped. This is different from NoImplicitAny in that it would not require types on everything. Further it would be a stronger guarantee than NoImplicitAny for the code that has types; use of *explicit* `any` defeats many guarantees of the typesystem.

In Other words, I'm proposing that that an `any` only be allowed to substitute for other types with a cast: `let f : DefinedThing = <DefinedThing>anyVar;` when a flag is given, such as `NoImplicitCastAny`.

In the language of C and C++, this is like saying I would like `any` to behave like C++'s void* rather than like C's void*.

**Comment by mhegazy** [comment]
`noImplicitAny` pushes the checks to production sites, this proposal wants it on consumption sites. each has its own merits.

One thing that `noImplicitAny` asserts, and would not be satisfied by this proposal, is that the compilation unit is safe for refactoring, as all declarations are strongly typed with the exception of clearly marked explicit any. 

We will need a detailed proposal for this. What type relations ships are affected by this switch, and in what way. details of evaluating type casts, and how to avoid reporting assignability errors in non-user code..

**Comment by DanielRosenwasser** [comment]
> Adding the type has given me no benefit at all.

I think you've gained the benefit of future typed code appropriately being checked.

I think our services layer addresses this scenario pretty well. We have tooling to let people find all references of `foo`, so you could track uses of `foo` down once you amend its signature.

**Comment by jargv** [comment]
@mhegazy true there are pros and cons of this vs noImplicityAny, I hadn't thought of them as mutually exclusive. I would expect to be able to enable zero, one, or both. 

I'm not sure what a "detailed proposal" would entail. Can you give me any more info?

**Comment by DanielRosenwasser** [comment]
@jargv you can check out our wiki page on [writing good design proposals](https://github.com/Microsoft/TypeScript/wiki/Writing-Good-Design-Proposals).

**Comment by jargv** [comment]
@DanielRosenwasser thanks, I'll give that a look.

**Comment by jargv** [comment]
I thought about this a bit more and realized that my assumptions about it were a bit off. 

I had assumed that given an untyped codebase, this proposal would allow me to add types to a single method and have the compiler signal any code which violates the new type constraints. The problem is that in an untyped codebase, the call sites are untyped, and so the compiler doesn't even know which method to check types against because the target of the method is an `any`. Closing.
